### DISCUSSION TITLE
Add limits / timeout to tasks

### DISCUSSION THREAD

**Discussion by donovanm21** [gameserverapp/Platform#154] [Suggestions]
Hi There,

Is it possible to have an option under the Scheduled or Triggered tasks to "run now"?

Background - I've put some shop packs together for players to have the ability to buy better multipliers for a specific time period on a specific map at a time. Currently I've add a shop pack that enables a disabled task, the task is to change the config file on a particular map and force the dynamic update to adjust the harvest rates from 2.0 to 2.5 as a 25% increase in the rate. 

The task itself has a queue that will after 58mins disable the task. 

The problem is the closest option I have at the moment is, set the minute of the hour option so run at min 1 (2:01) and disable 58 mins later (2:59) so it does not run again at 3:01. Issue is it will some times enable and run the task and other times it will only enable the task and wait till the 1 min is reached to run it. 

I've purchased the pack at say 2:25 and it ran at 2:26 which the enables the task but it's already run so now it will run twice due to the overlap.. 

Having it run "now" I can simply specify how long after the main task starts I want to have it disable, end or revert the main task action like in my case reverting the dynamic config to the standard rates. This way a player can buy and enable it at random times and have it active for a particular timeframe and it will revert.

Would be nice to have a "delay" option as well as part of the conditional logic to set a defined delay.

Add on to this option would be to check and prevent a purchase if the config file is already set, so a if / then conditional that will do a config check and reject the purchase based on the fact that it is already active. Almost like a "status" check on the current server environment.

Thanks in advance.
Donovan | yGaming.

**Comment by mvaessen** [comment]
Hi, can you add some screenshot for context? I'm not sure I fully understand your suggestion.

There are Delay options available for tasks, which can be configured for a specific amount of time. This is called "Queue / Schedule". You can find all blocks here: https://help.gameserverapp.com/article/229-task-delivery-builder

**Reply by donovanm21** [reply]
Hi Max, 

Thanks for the response. 

I've used the delay to disable the task once it's run after 58mins to prevent it from running again. 

Ideally what I would like to have is the ability to just run / enable the task once it is purchased and use the delay to disable it again with out having to specify a particular time when the task runs which is the issue and causes duplicate task to run.

I've attached screenshots with the current task and logic
![Screenshot 2022-04-11 at 11 35 34](https://user-images.githubusercontent.com/4052191/162711836-c2304dad-45f2-4dd7-a9a2-cbd14ebd9807.png)
![Screenshot 2022-04-11 at 11 35 43](https://user-images.githubusercontent.com/4052191/162711846-9be7de7c-c9e5-4e66-8ee5-2a07365f3e54.png)

**Comment by mvaessen** [comment]
Hi, thanks for the extra info. That helps better understand what you're trying to achieve.

I think the best way to mitigate this issue, is by adding timeouts / limits. I've discussed this with other customers in the past, so there are more people interested in a similar solution. Here is a link to a discord message that is about a similar suggestion: https://discord.com/channels/315049169071112193/685243791166537730/934463444676120716

Basically a timeout / limit would be something like a condition, where you can limit the number of runs. This would eliminate the need to disable tasks. It will simply skip anything inside the `timeout / limit` block.

I'm curious whether more people are interested in something like this.

**Reply by mvaessen** [reply]
Added additional suggestion #166

**Reply by donovanm21** [reply]
No problem, I am sure more people would be if they plan on using the shop to apply some updates to configuration and do certain tasks on the servers. Running a command is easy because it simply runs the command but running something for a certain period of time and then follow more logic become the challenge. 

Will have a look at the post and timeout / limit block to see if I can use that in the meanwhile to prevent to task from running more than once at a time. Fearful of creating a infinite loop.. :)

**Reply by mvaessen** [reply]
The timeout / limit block has not been added yet, but it was something we discussed.

You won't have to worry about infinite loops. GSA has protection in place to prevent this from happening. 👍

**Reply by donovanm21** [reply]
Awesome, will have a look at the discussion in the meanwhile.

Good to know on the protection side. Only have to experience a locked up system once to make sure you check code 5 times

**Comment by mvaessen** [comment]
Added to #517

**Comment by mvaessen** [comment]
Hi, while not exactly implemented as suggested are the options available on the dashboard to do the same thing.
